nilearn/regions/rena_clustering.py

```diff
@@ -153,6 +154,126 @@ def _make_edges_and_weights(X, mask_img):
+        Unmasked edges of the mesh.
+
+    Returns
+    -------
+    weights : ndarray
+        Weights corresponding to all edges in the mask.
+        shape: (n_edges,).
+
+    """
+    n_samples, _ = X.shape
+    shape = mask.shape
+
+    data = np.empty((shape[0], n_samples))
+    for sample in range(n_samples):
+        data[:, sample] = unmask_from_to_3d_array(X[sample].copy(), mask)
+
+    data_i = data[edges[0]]
+    data_j = data[edges[1]]
+    weights = np.sum((data_i - data_j) ** 2, axis=-1).ravel()
+
+    return weights
+
+
+def _make_edges_surface(faces, mask):
+    """Create the edges set: Returns a list of edges for a surface mesh.
+
+    Parameters
+    ----------
+    faces : ndarray
+        The vertex indices corresponding the mesh triangles.
+
+    mask : boolean
+        True if the edge is contained in the mask, False otherwise.
+
+    Returns
+    -------
+    edges : ndarray
+        Edges corresponding to the image with shape: (2, n_edges).
+
+    edges_masked : ndarray
+        Edges corresponding to the mask with shape: (1, n_edges).
+
+    """
+    mesh_edges = set()
+    for face in faces:
+        for i in range(len(face)):
+            edge = tuple(sorted([face[i], face[(i + 1) % len(face)]]))
+            mesh_edges.add(edge)
+    edges = np.array(list(mesh_edges))
+    false_indices = np.where(~mask)[0]
+    edges_masked = ~np.isin(edges, false_indices).any(axis=1)
+
+    return edges.T, edges_masked
+
+
+def _make_edges_and_weights_surface(X, mask_img):
+    """Compute the weights to all edges in the mask.
+
+    Parameters
+    ----------
+    X : ndarray, shape = [n_samples, n_features]
+        Training data.
+
+    mask_img : SurfaceImage object
+        Object used for masking the data.
+
+    Returns
+    -------
+    edges : dict[str, np.array]
+        Array containing edges of mesh
+
+    weights : dict[str, np.array]
+        Weights corresponding to all edges in the mask.
+        shape: (n_edges,).
+
+    """
+    weights = {}
+    edges = {}
+    len_previous_mask = 0
+    for part in mask_img.mesh.parts:
+        face_part = mask_img.mesh.parts[part].faces
+        mask_part = mask_img.data.parts[part]
+
+        edges_unmasked, edges_mask = _make_edges_surface(face_part, mask_part)
+
+        idxs = np.array(range(mask_part.sum())) + len_previous_mask
+        weights_unmasked = _compute_weights_surface(
+            X[:, idxs], mask_part.astype("bool"), edges_unmasked
+        )
+        # Apply mask to edges and weights
+        weights[part] = np.copy(weights_unmasked[edges_mask])
+        edges_ = np.copy(edges_unmasked[:, edges_mask])
+
+        # Reorder the indices of the graph
+        max_index = edges_.max()
+        order = np.searchsorted(
+            np.unique(edges_.ravel()), np.arange(max_index + 1)
+        )
+        # increasing the order by the number of vertices in the previous mask
+        # to avoid overlapping indices
+        order += len_previous_mask
```

Comment:
@bthirion I made a change here because IIUC this comment https://github.com/nilearn/nilearn/pull/4205#discussion_r1461818983, we should not have any overlapping vertex indices between the left and right parts.

This wasn't the case before, see commit https://github.com/nilearn/nilearn/pull/4577/commits/ce895c53358106f5f490011584a4fd6f6d0a0584.

So just making sure I understand this correctly. Please LMK. Thanks!